stringdtype/stringdtype/src/umath.c

```diff
@@ -73,36 +137,74 @@ init_equal_ufunc(PyObject *numpy)
     /*
      * Initialize spec for equality
      */
-    PyArray_DTypeMeta **eq_dtypes = malloc(3 * sizeof(PyArray_DTypeMeta *));
-    eq_dtypes[0] = &StringDType;
-    eq_dtypes[1] = &StringDType;
-    eq_dtypes[2] = &PyArray_BoolDType;
+    PyArray_DTypeMeta *eq_dtypes[3] = {&StringDType, &StringDType,
+                                       &PyArray_BoolDType};
 
     static PyType_Slot eq_slots[] = {
             {NPY_METH_resolve_descriptors, &string_equal_resolve_descriptors},
             {NPY_METH_strided_loop, &string_equal_strided_loop},
             {0, NULL}};
 
-    PyArrayMethod_Spec *EqualSpec = malloc(sizeof(PyArrayMethod_Spec));
+    PyArrayMethod_Spec EqualSpec = {
+            .name = "string_equal",
+            .nin = 2,
+            .nout = 1,
+            .casting = NPY_NO_CASTING,
+            .flags = 0,
+            .dtypes = eq_dtypes,
+            .slots = eq_slots,
+    };
+
+    if (PyUFunc_AddLoopFromSpec(equal, &EqualSpec) < 0) {
+        Py_DECREF(equal);
+        return -1;
+    }
 
-    EqualSpec->name = equal_name;
-    EqualSpec->nin = 2;
-    EqualSpec->nout = 1;
-    EqualSpec->casting = NPY_SAFE_CASTING;
-    EqualSpec->flags = 0;
-    EqualSpec->dtypes = eq_dtypes;
-    EqualSpec->slots = eq_slots;
+    /*
+     * This might interfere with NumPy at this time.
+     */
+    PyObject *promoter_capsule1 = PyCapsule_New(
+            (void *)&default_ufunc_promoter, "numpy._ufunc_promoter", NULL);
+    if (promoter_capsule1 == NULL) {
+        return -1;
+    }
 
-    if (PyUFunc_AddLoopFromSpec(equal, EqualSpec) < 0) {
-        Py_DECREF(equal);
-        free(eq_dtypes);
-        free(EqualSpec);
+    PyObject *DTypes1 = PyTuple_Pack(3, &StringDType, &PyArray_UnicodeDType,
+                                     &PyArrayDescr_Type);
+    if (DTypes1 == 0) {
+        Py_DECREF(promoter_capsule1);
+        return -1;
+    }
+
+    if (PyUFunc_AddPromoter(equal, DTypes1, promoter_capsule1) < 0) {
+        Py_DECREF(promoter_capsule1);
```

Comment:
You can use the same promoter twice, nothing is prevnting you from that.  Otherwise LGTM (and happy either way).